### PR TITLE
working out the hallway example, currently stuck on something related…

### PR DESCRIPTION
**Pull request by pstjohn**
… to tf datasets

### CODE REVIEW

File: rlmolecule/hallway/hallway_state.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return type(self) == type(other) and \
+               self._position == other._position and \
+               self._forced_terminal == other._forced_terminal 
+
+    def hash(self) -> int:
+        return hash(self.__repr__()) ^ (13 * self._forced_terminal)
+
+    def get_next_actions(self) -> Sequence['HallwayState']:
+        result = []
+        if not self._forced_terminal:
```
Comment (pstjohn) [inline rlmolecule/hallway/hallway_state.py:30]: you certainly don't have to use the `forced_terminal` motif, for molecules, that was the only way I could figure out to have 'endpoints' at any node.

File: rlmolecule/hallway/hallway_problem.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        :return:
+        """
+
+        def get_policy_inputs_tf(parent, reward, visit_probabilities,
+                                 problem: HallwayAlphaZeroProblem) -> {}:
+            position = tf.py_function(
+                problem._get_network_inputs_from_serialized_parent, inp=[parent],
+                Tout=[tf.int64])
+            print(position)
+            position.set_shape([None, None])
```
Comment (pstjohn) [inline rlmolecule/hallway/hallway_problem.py:121]: is this where you're getting the error? I would think position would be a [None, 1] or [None] shaped vector here?

Yeah, looks like you use it as [None, 1] later as a neural network input. You might want to make sure position coming out of `_get_network_inputs_from_serialized_parent` is a 2D numpy array

File: rlmolecule/hallway/policy/model.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+def policy_model(hallway_size: int,
+                 features: int = 8,
+                 hidden_layers: int = 3,
+                 hidden_dim: int = 16) -> tf.keras.Model:
+
+    position_class = layers.Input([None, 1], dtype=tf.int64, name="position")
+    x = layers.Embedding(hallway_size, features, name="embedding")(position_class)
+    
+    for layer in range(hidden_layers):
+        x = layers.Dense(hidden_dim)(x)
```
Comment (pstjohn) [inline rlmolecule/hallway/policy/model.py:16]: you'd probably want an activation layer (and maybe batchnorm) between these Dense layers. Although for this simple problem it likely doesn't matter. But as written, these dense layers would collapse to functionally a single matrix multiplication

File: rlmolecule/hallway/policy/model.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        flat_values_logits, flat_prior_logits = self.policy_model([positions_flat])
+
+        # We put the parent node first in our batch inputs, so this slices
+        # the value prediction for the parent
+        value_preds = tf.reshape(flat_values_logits, [batch_size, max_actions_per_node, -1])[:, 0, 0]
+
+        # Next we get a mask to see where we have valid actions and replace priors for
+        # invalid actions with negative infinity (these get zeroed out after softmax).
+        # We also only return prior_logits for the child nodes (not the first entry)
+        action_mask = tf.reduce_any(tf.not_equal(positions, 0), axis=-1)  # zero is the padding element
```
Comment (pstjohn) [inline rlmolecule/hallway/policy/model.py:64]: note to myself to find a way to use keras masking to do this

**Review by pstjohn — COMMENTED** [review]
Looks good, when I get a chance I'm happy to clone and see if I can figure out the tf.dataset issue. In general, I think we wouldn't want a separate hallway tree in rlmolecule/, but it would live instead in examples/. But in the meantime this is fine. ultimately it would be nice to have the example live in a single python file (or just a couple) rather than having to implement an entire directory...

File: rlmolecule/hallway/hallway_state.py (hunk 1 above)
Comment (davebiagioni) [reply rlmolecule/hallway/hallway_state.py:30]: for sure, was just trying to do minimal changes to get this to work.  we do need to set a terminal state but there isn't a notion of "forcing" it for the hallway example

File: rlmolecule/hallway/policy/model.py (hunk 3 above)
Comment (davebiagioni) [reply rlmolecule/hallway/policy/model.py:16]: oh wow, how did i forget that... yeah i think i can just add `activation="relu"` in the Dense init

File: rlmolecule/hallway/hallway_problem.py (hunk 2 above)
Comment (davebiagioni) [reply rlmolecule/hallway/hallway_problem.py:121]: makes sense, yeah i was working backwards through this.  i had [None, 1] at one point but this also failed, probably for the reason you mentioned
